### DISCUSSION TITLE
How can test and real components have different dependency graphs?

### DISCUSSION THREAD

**Discussion by suyash192** [evant/kotlin-inject#330] [Q&A]
Here's a modified version of the component inheritance example from README.md.

```Kotlin
@NetworkScope
abstract class NetworkComponent {
    @NetworkScope
    @Provides
    abstract fun api(): Api
}

@Component
abstract class RealNetworkComponent : NetworkComponent() {
    override fun api(depA: DepA): Api = RealApi(depA)
}

@Component
abstract class TestNetworkComponent : NetworkComponent() {
    override fun api(depB: DepB): Api = FakeApi(depB)
}

@Component abstract class AppComponent(@Component val network: NetworkComponent)
```

Notice that the `api()` functions in `RealNetworkComponent` and `TestNetworkComponent` have different dependencies (`DepA` and `DepB`). This will obviously not compile. But in real world situations test classes and prod classes can have different dependencies. How do we achieve this using kotlin-inject?

One solution would be to add all the dependencies to the `NetworkComponent.api` function signature.

```Kotlin
@NetworkScope
@Provides
abstract fun api(depA: DepA, depB: DepB): Api
```

and then the sub classes can simply ignore the dependency that they don't need. This would work but it feels like a hacky workaround rather than a proper solution. And there might be situations where `DepA` will be present in main src while `DepB` will be present in test src.

So what would be the proper solution to this?

**Comment by evant** [comment]
Does https://github.com/evant/kotlin-inject/blob/main/docs/testing.md answer your question? The README sample was written before that and probably could be updated with a more realistic usecase.

**Comment by suyash192** [comment]
It does not. These are the classes in testing.md, they have no dependencies, unlike the example I've given above.

```Kotlin
    @get:Provides val accountService: AccountService = FakeAccountService(),
    @get:Provides val userService: UserService = FakeUserService(),
```

**Comment by evant** [comment]
You want to inject into your test fakes? There's nothing stopping you from declaring extra provides in TestApplicationComponent.

**Comment by suyash192** [comment]
@evant I'm still unable to understand. `AppComponent` accepts `NetworkComponent` in its constructor as it needs the `Api` dependency declared in it. It does not accept `TestNetworkComponent` and knows nothing about the additional functions in it. If I add extra provides functions in `TestNetworkComponent` then they will not be used by `AppComponent` as it does not know about the additional functions in `TestNetworkComponent`.

```Kotlin
@Component abstract class AppComponent(@Component val network: NetworkComponent)
```
